man/query.1
.\" Automatically generated by Pandoc 2.6
.\"
.TH "Collins Golang CLI" "" "February 12, 2019" "" ""
.hy
.SH Synopsis
.PP
\f[C]collins query\f[R] options [hostname]\&...
.SH Description
.PP
Collins query provides a wrapper round the collins find endpoint.
Most input is just taken from passed in flags and converted into a
collins query using CQL.
Multiple output formats are supported such as Table, YAML, JSON, and
Links.
Table which is the default output format can be used to pipe the output
into other commands such as set, power, and provision.
.SS Query Examples
.PP
A basic example to query a specific tag and additionally tell us the ip
of the asset would be:
.IP
.nf
\f[C]
collins query -t M000001 -x ip_address
\f[R]
.fi
.SH Options
.SS Query options
.TP
.B \f[C]-t\f[R] \f[I]VALUES\f[R], \f[C]--tag\f[R] \f[I]VALUES\f[R]
Specify tag names to return they must be the full tag and in the case
that multiple are specified they must be separated by commas.
.TP
.B \f[C]-Z\f[R], \f[C]--remote-lookup\f[R]
Remote lookup will utilize collins remote lookup feature to query not
only the current collins instances but other instances it has been
linked with as well.
.TP
.B \f[C]-T\f[R] \f[I]VALUES\f[R], \f[C]--type\f[R] \f[I]VALUES\f[R]
Specify the type of asset that should be returned.
Common uses for this might be for selecting all power switches or server
nodes.
When multiple values are specified they must be separated by commas.
.RS
.IP
.nf
\f[C]
-T SERVER_NODE
--type POWER_SWITCH,SERVER_NODE
\f[R]
.fi
.RE
.TP
.B \f[C]-n\f[R] \f[I]VALUES\f[R], \f[C]--nodeclass\f[R] \f[I]VALUES\f[R]
Specify the nodeclass to query.
When multiple values are specified they must be separated by commas.
By default this does a wildcard match at the start and end of the string
so a value of \f[C]dev\f[R] would also match \f[C]clusterdevnode\f[R].
.TP
.B \f[C]-p\f[R] \f[I]VALUES\f[R], \f[C]--pool\f[R] \f[I]VALUES\f[R]
Specify the pool to query.
When multiple values are specified they must be separated by commas.
By default this does a wildcard match at the start and end of the string
so a value of \f[C]prod\f[R] would also match \f[C]preprod01\f[R].
.TP
.B \f[C]-r\f[R] \f[I]VALUES\f[R], \f[C]--role\f[R] \f[I]VALUES\f[R]
Specify the primary role to query.
When multiple values are specified they must be separated by commas.
By default this does a wildcard match at the start and end of the string
so a value of \f[C]MYSQL\f[R] would also match \f[C]MYSQL-DEV\f[R].
.TP
.B \f[C]-R\f[R] \f[I]VALUES\f[R], \f[C]--secondary-role\f[R] \f[I]VALUES\f[R]
Specify the secondary role to query.
When multiple values are specified they must be separated by commas.
By default this does a wildcard match at the start and end of the string
so a value of \f[C]MASTER\f[R] would also match \f[C]MESOS-MASTER\f[R].
.TP
.B \f[C]-i\f[R] \f[I]VALUES\f[R], \f[C]--ip-address\f[R] \f[I]VALUES\f[R]
Specify the ip address to query.
When multiple values are specified they must be separated by commas.
This does not do wildcard matching.
.TP
.B \f[C]-S\f[R] \f[I]VALUE\f[R], \f[C]--status\f[R] \f[I]VALUE\f[R]
Query only for assets that are in a given state.
Status takes a argument in the form of status:state.
An example would be \f[C]Allocated:Running\f[R].
State is however not required and you can just query for
\f[C]Allocated\f[R] as well.
.TP
.B \f[C]-a\f[R] \f[I]VALUE\f[R], \f[C]--attribute\f[R] \f[I]VALUE\f[R]
This is the bread and butter you can query for any key value pair you
want using this flag.
It can be specified multiple times and looks like \f[C]key:value\f[R].
.TP
.B \f[C]-o\f[R] \f[I]VALUE\f[R], \f[C]--operation\f[R] \f[I]VALUE\f[R]
Specify operation to join all flags with.
By default AND is used.
\f[I]VALUE\f[R] can be:
.RS
.IP \[bu] 2
\f[C]AND\f[R]
.IP \[bu] 2
\f[C]OR\f[R]
.RE
.TP
.B \f[C]-s\f[R] \f[I]VALUE\f[R], \f[C]--size\f[R] \f[I]VALUE\f[R]
Set the number of results to query for at once.
This is by default set to be 100.
If you are running this from some place where network latency is a huge
issue this may be useful to increase.
.TP
.B \f[C]--limit\f[R] \f[I]VALUE\f[R]
Limit the total number of assets that will be returned for the query.
By default everything will be returned.
.SS Format options
.TP
.B \f[C]-H\f[R], \f[C]--show-header\f[R]
If you are rendering output in table format show the header above each
column.
.TP
.B \f[C]-c\f[R], \f[I]VALUES\f[R], \f[C]--columns\f[R] \f[I]VALUES\f[R]
Specify a list of all of the attributes that should be shown for table
formatting.
By default we first check if the value is a custom one that we know
about and need to compute such as total number of CPUs.
If it\[cq]s not a special column we check if it\[cq]s in attributes when
outputting it.
The following special values are available.
.RS
.PP
By default tag, hostname, nodeclass, status, pool, primary_role, and
seconadary_role are returned.
.IP \[bu] 2
tag
.IP \[bu] 2
status
.IP \[bu] 2
state
.IP \[bu] 2
classification
.IP \[bu] 2
ip_address
.IP \[bu] 2
ipmi_address
.IP \[bu] 2
cpu_cores
.IP \[bu] 2
cpu_threads
.IP \[bu] 2
cpu_speed_ghz
.IP \[bu] 2
cpu_description
.IP \[bu] 2
gpu_description
.IP \[bu] 2
cpu_product
.IP \[bu] 2
gpu_product
.IP \[bu] 2
cpu_vendor
.IP \[bu] 2
gpu_vendor
.IP \[bu] 2
memory_size_bytes
.IP \[bu] 2
memory_size_total
.IP \[bu] 2
memory_description
.IP \[bu] 2
memory_banks_total
.RE
.TP
.B \f[C]-x\f[R], \f[I]VALUES\f[R], \f[C]--extra-columns\f[R] \f[I]VALUES\f[R]
Instead of overwritting the default output columns with \f[C]-c\f[R] you
can append to them using \f[C]-x\f[R].
The same documentaion applies to this.
.TP
.B \f[C]-f\f[R] \f[I]VALUE\f[R], \f[C]--field-separator\f[R] \f[I]VALUE\f[R]
This sets the delimiter that all columns are seperated by.
By default all columns are seperated with a tab.
.SS Robot options
.TP
.B \f[C]-l\f[R], \f[C]--link\f[R]
Output the assets with a link to them in the web UI.
.TP
.B \f[C]-j\f[R], \f[C]--json\f[R]
Output the assets as a JSON array.
.TP
.B \f[C]-y\f[R], \f[C]--yaml\f[R]
Output the assets as a YAML array.
.SH AUTHORS
Michael Schuett.


cmds/query.go
package cmds

import (
	"os"
	"strings"

	cli "github.com/urfave/cli"
	collins "gopkg.in/tumblr/go-collins.v0/collins"
)

func QuerySubcommand() cli.Command {
	return cli.Command{
		Name:    "query",
		Aliases: []string{"find"},
		Usage:   "Search for assets in Collins",
		Flags: []cli.Flag{
			cli.StringFlag{
				Name:     "t, tag",
				Usage:    "Assets with tag[s] value[,...]",
				Category: "Query options",
			},
			cli.BoolFlag{
				Name:     "Z, remote-lookup",
				Usage:    "Query remote datacenters for asset",
				Category: "Query options",
			},
			cli.StringFlag{
				Name:     "T, type",
				Usage:    "Only show asset with type value",
				Category: "Query options",
			},
			cli.StringFlag{
				Name:     "n, nodeclass",
				Usage:    "Assets in nodeclass value[,...]",
				Category: "Query options",
			},
			cli.StringFlag{
				Name:     "p, pool",
				Usage:    "Assets in pool value[,...]",
				Category: "Query options",
			},
			cli.IntFlag{
				Name:     "s, size",
				Usage:    "Number of assets to return per page",
				Value:    100,
				Category: "Query options",
			},
			cli.IntFlag{
				Name:     "limit",
				Usage:    "Limit total results of assets",
				Category: "Query options",
			},
			cli.StringFlag{
				Name:     "r, role",
				Usage:    "Assets in primary role[s] value[,...]",
				Category: "Query options",
			},
			cli.StringFlag{
				Name:     "R, secondary-role",
				Usage:    "Assets in secondary role[s] value[,...]",
				Category: "Query options",
			},
			cli.StringFlag{
				Name:     "i, ip-address",
				Usage:    "Assets with IP address[es] value[,...]",
				Category: "Query options",
			},
			cli.StringFlag{
				Name:     "S, status",
				Usage:    "Asset status (and optional state after :)",
				Category: "Query options",
			},
			cli.StringSliceFlag{
				Name:     "a, attribute",
				Usage:    "Arbitrary attributes and values to match in query. : between key and value",
				Category: "Query options",
			},
			cli.StringFlag{
				Name:     "o, operation",
				Usage:    "Sets if your query will be joined with AND or OR",
				Value:    "AND",
				Category: "Query options",
			},
			cli.BoolFlag{
				Name:     "H, show-header",
				Usage:    "Show header fields in output",
				Category: "Table formatting",
			},
			cli.StringFlag{
				Name:     "c, columns",
				Usage:    "Attributes to output as columns, comma separated",
				Value:    "tag,hostname,nodeclass,status,pool,primary_role,secondary_role",
				Category: "Table formatting",
			},
			cli.StringFlag{
				Name:     "x, extra-columns",
				Usage:    "Show these columns in addition to the default columns, comma separated",
				Category: "Table formatting",
			},
			cli.StringFlag{
				Name:     "f, field-separator",
				Usage:    "Separator between columns in output",
				Value:    "\t",
				Category: "Table formatting",
			},
			cli.BoolFlag{
				Name:     "l, link",
				Usage:    "Output link to assets found in web UI",
				Category: "Robot formatting",
			},
			cli.BoolFlag{
				Name:     "j, json",
				Usage:    "Output results in JSON",
				Category: "Robot formatting",
			},
			cli.BoolFlag{
				Name:     "y, yaml",
				Usage:    "Output results in YAML",
				Category: "Robot formatting",
			},
		},
		Action: queryRunCommand,
	}
}

func queryBuildOptions(c *cli.Context, hostname string) collins.AssetFindOpts {
	opts := collins.AssetFindOpts{}

	if c.IsSet("status") {
		status := strings.Split(c.String("status"), ":")
		if len(status) == 2 {
			opts.State = status[1]
		}
		opts.Status = status[0]
	}

	if c.IsSet("type") {
		opts.Type = c.String("type")
	}

	if c.IsSet("remote-lookup") {
		opts.RemoteLookup = true
	}

	opts.Query = buildOptionsQuery(c, hostname)

	return opts
}

// Build a CQL query from the passed in flag taking into account comma seperated values.
func cqlQuery(c *cli.Context, flag string, field string) string {
	query := []string{}
	for _, val := range strings.Split(c.String(flag), ",") {
		query = append(query, "("+field+" = "+val+")")
	}

	// This isn't needed but keeps the resulting CQL from not looking so crazy
	if len(query) == 1 {
		return query[0]
	} else {
		return "(" + strings.Join(query, " OR ") + ")"
	}
}

// This is broke out of build options just for the sake of making testing easier
func buildOptionsQuery(c *cli.Context, hostname string) string {
	cql := []string{}
	// The go client isn't as friendly as the ruby one which is fine we will just
	// take everything else and convert it into CQL to talk to collins.
	if c.IsSet("tag") {
		cql = append(cql, cqlQuery(c, "tag", "TAG"))
	}

	if c.IsSet("nodeclass") {
		cql = append(cql, cqlQuery(c, "nodeclass", "NODECLASS"))
	}

	if c.IsSet("pool") {
		cql = append(cql, cqlQuery(c, "pool", "POOL"))
	}

	if c.IsSet("role") {
		cql = append(cql, cqlQuery(c, "role", "PRIMARY_ROLE"))
	}

	if c.IsSet("secondary-role") {
		cql = append(cql, cqlQuery(c, "secondary-role", "SECONDARY_ROLE"))
	}

	if c.IsSet("ip-address") {
		cql = append(cql, cqlQuery(c, "ip-address", "IP_ADDRESS"))
	}

	if hostname != "" {
		cql = append(cql, "(HOSTNAME = "+hostname+")")
	}

	if c.IsSet("attribute") || c.IsSet("a") {
		for _, attr := range c.StringSlice("attribute") {
			attrSplit := strings.SplitN(attr, ":", 2)
			if len(attrSplit) != 2 {
				logAndDie("--attribute and -a requires attribute:value, missing :value")
			}
			attrKey := strings.ToUpper(attrSplit[0])
			attrValue := strings.ToUpper(attrSplit[1])

			cql = append(cql, "("+attrKey+" = "+attrValue+")")
		}
	}

	operation := c.String("operation")
	if operation != "AND" && operation != "OR" {
		logAndDie("Operation (or o) flag may only be set to AND or OR")
	}

	return strings.Join(cql, " "+operation+" ")
}

// This uses a "trick" of using a map to create a unique list that
// we then turn into a slice before returning.
func queryGetColumns(c *cli.Context) []string {
	uniqueSet := UniqueOrderedSet{
		"tag",
		"hostname",
		"nodeclass",
		"status",
		"pool",
		"primary_role",
		"secondary_role",
	}

	if c.IsSet("attribute") || c.IsSet("a") {
		for _, attr := range c.StringSlice("attribute") {
			attrSplit := strings.SplitN(attr, ":", 2)
			if len(attrSplit) != 2 {
				logAndDie("--attribute and -a requires attribute:value, missing :value")
			}
			uniqueSet = uniqueSet.Add(attrSplit[0])
		}
	}

	if c.IsSet("columns") {
		uniqueSet = strings.Split(c.String("columns"), ",")
	}

	if c.IsSet("extra-columns") {
		extras := strings.Split(c.String("extra-columns"), ",")
		for _, column := range extras {
			uniqueSet = uniqueSet.Add(column)
		}
	}

	return uniqueSet
}

func getOutputFormat(c *cli.Context) string {
	switch {
	case c.IsSet("link"):
		return "link"
	case c.IsSet("json"):
		return "json"
	case c.IsSet("yaml"):
		return "yaml"
	default:
		return "table"
	}
}

func queryRunCommand(c *cli.Context) error {
	// Don't run if nothing is passed into the command
	if len(os.Args) == 2 {
		logAndDie("See --help for collins query usage")
	}

	// If the user passes in an argument we treat it as a
	// hostname and pass it along it overwrites hostname
	// in the case you set it as an attribute
	hostname := ""
	if c.NArg() > 0 {
		hostname = c.Args().Get(0)
	}

	client := getCollinsClient(c)
	opts := queryBuildOptions(c, hostname)
	size := c.Int("size")

	// Kinda hacky but if limit is set we just set
	// that as the page size and break after the first
	// call to get assets.
	if c.IsSet("limit") {
		size = c.Int("limit")
	}

	opts.PageOpts = collins.PageOpts{
		Size: size,
	}

	var allAssets []collins.Asset
	for {
		assets, resp, err := client.Assets.Find(&opts)

		if err != nil {
			logAndDie(err.Error())
		}

		allAssets = append(allAssets, assets...)

		// Limit was set break out of the loop now
		if c.IsSet("limit") {
			break
		}

		if resp.NextPage == resp.CurrentPage { // No more pages
			break
		} else { // Fetch next page
			opts.PageOpts.Page++
		}
	}

	columns := queryGetColumns(c)
	format := getOutputFormat(c)
	showHeaders := c.Bool("show-header")
	formatAssets(format, c.String("field-separator"), showHeaders, client.BaseURL.String(), c.Bool("remote-lookup"), columns, allAssets)

	return nil
}
